code/game/objects/items/RCD.dm

```diff
@@ -670,81 +533,136 @@ GLOBAL_VAR_INIT(icon_holographic_window, init_holographic_window())
+	return ui_always_active ? GLOB.always_state : ..()
+
+/obj/item/construction/rcd/ui_static_data(mob/user)
+	return airlock_electronics.ui_static_data(user)
+
+/obj/item/construction/rcd/ui_data(mob/user)
+	var/list/data = ..(user)
+
+	data["space"] = TRUE
+	//main categories
+	data["selected_root"] = root_category
+	data["root_categories"] = list()
+	for(var/category in root_categories)
+		data["root_categories"] += category
+	data["root_categories"] += "AirLock Access"
+
+	//create the category list
+	data["selected_category"] = design_category
+	data["selected_design"] = design_title
+	data["categories"] = list()
+
+	for(var/list/sub_category as anything in root_categories[root_category])
+		//skip category if upgrades were not installed for these
+		if(sub_category == "Machines" && !(upgrade & RCD_UPGRADE_FRAMES))
+			continue
+		if(sub_category == "Furniture" && !(upgrade & RCD_UPGRADE_FURNISHING))
+			continue
+
+		var/list/designs = list() //initialize all designs under this category
+		var/list/target_category =  root_categories[root_category][sub_category]
+		for(var/i in 1 to target_category.len)
+			var/list/design = target_category[i]
+
+			//get/infer icon_state based on category
+			var/icon
+			if(sub_category == "Solid AirLocks")
+				icon = "[design["title"]]"
+			else if(sub_category == "Glass AirLocks")
+				icon = "[design["title"]]-Glass"
+			else
+				icon = design["icon"]
+			icon = replacetext(icon," ","-") //css class names cannot have spaces in them
+
+			designs += list(list("design_id" = i, "title" = design["title"], "icon" = icon))
+		data["categories"] += list(list("cat_name" = sub_category, "designs" = designs))
+
+	//merge airlock_electronics ui data with this
+	var/list/airlock_data = airlock_electronics.ui_data(user)
+	for(var/key in airlock_data)
+		data[key] = airlock_data[key]
+
+	return data
+
+/obj/item/construction/rcd/ui_act(action, params)
+	..()
+	if(.)
 		return
-	switch(choice)
-		if("Floors & Walls")
-			construction_mode = RCD_FLOORWALL
-		if("Airlock")
-			construction_mode = RCD_AIRLOCK
-		if("Grilles & Windows")
-			construction_mode = RCD_WINDOWGRILLE
-		if("Machine Frames")
-			construction_mode = RCD_MACHINE
-		if("Furnishing")
-			construction_mode = RCD_FURNISHING
-		if("Computer Frames")
-			construction_mode = RCD_COMPUTER
-			change_computer_dir(user)
-			return
-		if("Change Access")
-			airlock_electronics.ui_interact(user)
-			return
-		if("Change Airlock Type")
-			change_airlock_setting(user)
-			return
-		if("Change Window Glass")
-			toggle_window_glass(user)
-			return
-		if("Change Window Size")
-			toggle_window_size(user)
-			return
-		if("Change Furnishing Type")
-			change_furnishing_type(user)
-			return
-		if("Silo Link")
-			toggle_silo_link(user)
-			return
+
+	switch(action)
+		if("root_category")
+			var/new_root = params["root_category"]
+			if(root_categories[new_root] != null || new_root == "AirLock Access") //is a valid category
```

Comment:
Why are we referencing specific roots here? What is this for?

Reply:
Airlock access is an special exception because it does not have items to display but it shows the Airlock electronics access panel UI.
So I need an exceptional check for this because it's not defined in the static list root_categories

Reply:
Never mind fixed this